src/BiDirectionalLinkedListImp.h

```diff
@@ -0,0 +1,152 @@
+    }
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::findNodeByValue(const Type &valueToFind) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            return current;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::deleteNodeByValue(const Type &valueToFind) {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            remove(current);
+
+            break;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printList(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        os << current->getValue() << " ";
+    }
+
+    os << std::endl;
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printListReverse(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->last; current != nullptr;
+         current = current->getPrevious().lock()) {
+        os << current->getValue() << " ";
+    }
+
+    os << std::endl;
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>>
+BiDirectionalLinkedList<Type>::insertNodeAfter(shared_ptr<BiDirectionalNode<Type>> nodeToInsertAfter,
+                                               shared_ptr<BiDirectionalNode<Type>> newNode) {
+    // If list is empty we insert it to the the head of the list
+    if (getFirst() == nullptr) {
+        assert(nodeToInsertAfter == nullptr);
+
+        this->first = newNode;
+        this->last = newNode;
+
+        return getFirst();
+    }
+
+    newNode->setPrevious(nodeToInsertAfter);
+
+    if (nodeToInsertAfter != nullptr) {
+        newNode->setNext(nodeToInsertAfter->getNext());
+        nodeToInsertAfter->setNext(newNode);
+    } else {
+        newNode->setNext(this->first);
+        this->first = newNode;
+    }
+
+    if (newNode->getNext() != nullptr) {
+        newNode->getNext()->setPrevious(newNode);
+    }
+
+    if (nodeToInsertAfter == getLast()) {
+        this->last = newNode;
+    }
+
+    return newNode;
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::remove(shared_ptr<BiDirectionalNode<Type>> nodeToDelete) {
+    assert(nodeToDelete != nullptr);
+
+    if (nodeToDelete->getNext() != nullptr) {
+        nodeToDelete->getNext()->setPrevious(nodeToDelete->getPrevious().lock());
+    }
+
+    if (nodeToDelete->getPrevious().lock() != nullptr) {
+        nodeToDelete->getPrevious().lock()->setNext(nodeToDelete->getNext());
+    }
+
+    if (nodeToDelete == getFirst()) {
+        this->first = this->first->getNext();
+    }
+
+    if (nodeToDelete == getLast()) {
+        this->last = nodeToDelete->getPrevious().lock();
+    }
+}
```

Comment:
ok i don't think i understand what happens in this function so maybe my comments are redundant.
i'd like to talk to you about this function 😟